### ISSUE TITLE
Next release WhereYouGo

### ISSUE THREAD

**Issue by Lineflyer** [cgeo/WhereYouGo#251]
I have seen some (mostly refactoring) activity ongoing for WhereYouGo.

Please tell me (by posting to this issue) if and when it would be desirable to release a new version?

Additionally I have just opened two issues, for which it might make sense to issue a release once they are fixed:
#249 #250

**Comment by SchoolGuy** [comment]
I believe that when I added the button to the settings to check the user credentials we could release a new version. To add that button my current Settings refactoring PRs needs to be well tested and merged. The button afterwards can be added very quickly. Thus I think end of February can be a good idea in my eyes.

**Comment by Lineflyer** [comment]
@SchoolGuy Any further activity planned from your side on the PR you mentioned above?
Indeed having a possibility to check the login would be good .

**Comment by SchoolGuy** [comment]
@Lineflyer We are waiting on @bekuno to review it. It is finished.

**Comment by Lineflyer** [comment]
Ah, sorry. I looked at #210 while you are probably refering to #240...

**Comment by SchoolGuy** [comment]
@Lineflyer Correct. But since #240 is a requirement for #210, it doesn't make a difference.

**Comment by Lineflyer** [comment]
AFAICS #240 is merged. What about #210?

I am sorry, that I don't follow this repo as close as I should....but time is limited.
#210 would be interesting to include in a potential next release.

**Comment by SchoolGuy** [comment]
@Lineflyer #210 is being worked on right now. So will be ready probably by end of the week if everything goes well.

**Comment by Lineflyer** [comment]
@SchoolGuy Very cool :)

**Comment by Lineflyer** [comment]
Due to some new problems (no location on Android 12) I created a project listing those issues, which are either regressions or mandatory changes to be done:
https://github.com/cgeo/WhereYouGo/projects/1

Summary:
- No GPS on Android 12
- No OSM Mapnik tiles (=default map not working anymore)
- No scoped storage
- Refactoring to target SDK30 or 31 needed
- Raise of minSDK needed to be able to update some dependencies

Sadly this does not mean, that it is anytime soon to happen. It purely depends on volunteers willing to step in and work on it.
Help needed.

**Comment by ztNFny** [comment]
Are we able to push a release to Play Store or are there any min/target-SDK or stuff requirements stopping us?

If not I'd suggest to push at least a small fix for the Android 12 and OSM tiles.

SAF is probably a bigger topic and will likely need also a new Mapsforge version - aka "big rewrite"

**Comment by Lineflyer** [comment]
We need to meet the requirement of targetSDK=30 to publish. I merged this change to master but did not do any extensive regression tests yet. It compiles and runs at least.
There is no requirement for the minSDK though.

So we would be able to publish right away from `master` but its...lets say....not well tested.
The problem is, that I can strive through the main functions and test on different emulators, but I don't have any setup to play a cartridge from my couch/on my emulator.

**Comment by ztNFny** [comment]
you can set the emulators location via telnet: "geo fix lon lat"

```
telnet localhost 5554
auth <key from file in your home dir>
geo fix 0 0
```

**Comment by ztNFny** [comment]
l'll try to run some tests on Android 11 and 12. Cant go lower than SDK 24

**Comment by Lineflyer** [comment]
I will do some more testing with nightly and emulator for `master` but won`t find the time to play cartridges in real life or emulated. 
If I don`t see any obstacles I suggest to push a new version (and evaluate from user feedback and Google reports).

Is there a risk that the OpenWIG code has some problems with the new targetSDK?? Anyone with insights?

**Comment by Lineflyer** [comment]
I tested the app up to best possibility. Not really in depth due to missing time and possibility to play cartridges.
I added a changelog on `master` to prepare version 2022.01.02.

If no one objects I will forward `release` and just push it tonight to Google Play.
There is some risk, that due to the new targetSDK something is not working as expected but I would still like to try to restore functionality for Android 12 and OSM online maps for the users.

**Comment by Lineflyer** [comment]
Just played a cartridge using mock locations and all went well on Android 11.
Just #329 is a regression...if anyone has some free time to look at that, it would be appreciated. I assume its no difficult bug.

Still planning to release tonight (at least on the beta channel...1456 testers).

**Comment by Lineflyer** [comment]
Tests (including the latest hotfix for #329) look OK or at least I am confident enough to push an update to the beta testers.
Will prepare now.

**Comment by Lineflyer** [comment]
While smoke-testing (in emulator) I found, that the app is not working on API16 and working on API19 but with some problems. Tests on API21 seemed to look normal though.

Therefore I decided to merge #273 in order not to deliver a non-working version to API 16-20 users.
A follow up issue regarding a possible legacy release has been created: #333 

Will continue checking and releasing now.

**Comment by Lineflyer** [comment]
WhereYouGo version 2022.01.02 has been pushed to beta users.

As we don't have a separate beta versioning, the plan would be to promote this version into normal production if beta test is normal (maybe next weekend or even earlier).

Tagging and uploading to Github ongoing now.

**Comment by Lineflyer** [comment]
It took a while but version 2022.01.02 is available to beta testers now. Tagging and uploading on Github also completed.

**Comment by Lineflyer** [comment]
FYI:
Some first feedback on social channels is positive for Android 12 functionality and general cartridge playing.

**Comment by sk-juroot** [comment]
Same positive feedback from Slovak geocaching group, people are just asking for SD card support and offline maps functionality. I will look at the requirements for this big refactor later, to see what is possible.

**Comment by ztNFny** [comment]
> Same positive feedback from Slovak geocaching group, people are just asking for SD card support 

https://github.com/cgeo/WhereYouGo/issues/309

> and offline maps functionality

https://github.com/cgeo/WhereYouGo/issues/73

The latest status on that (from c:geo dev meeting) is that it's not really possible to share the maps component between those two as c:geo just has too much "special" stuff in there. Possible routes would be
- keep WYG completly independent (high effort)
- integrate it into c:geo
- recommend Locus as main map option for WYG (lowest effort)

**Comment by bekuno** [comment]
Additional solution
- use c:geo as map provider in the same way as locus (maybe easier  to achieve than a complete new mapsforge solution)

**Comment by ztNFny** [comment]
> * use c:geo as map provider in the same way as locus (maybe easier  to achieve than a complete new mapsforge solution)

not sure about that one effort wise... WIGs require drawing shapes, something not currently supported by c:geo

But for any solution (but the Locus one) we need someone willing to spend a considerable amount of time

**Comment by sk-juroot** [comment]
Some cartridges are unplayable when Locus is used as a map provider - when they manipulate zones from timer/lua code, it breaks. (And this is only my personal experience from playing some local wig, I have not tested this functionality deeply.)

With all the special requirements for both c:geo and WhereYouGo, I would suggest (and also start looking into it) the hardest route - independent solution based on updated Mapsforge and following required components. In the end, they may share mapfiles and have some degree of integration, but should operate on their owns. As you say, c:geo have its own specialities, as WYG have.

**Comment by bekuno** [comment]
@sk-juroot The own mapsforge solution for WYG would real great.

**Comment by Lineflyer** [comment]
Reminder for myself:
Recent fixes have been merged to `master` but shall be included before going from beta to public.
This means, `master` needs to be merged to `release` another time before next beta/public.

**Comment by Lineflyer** [comment]
`release` has been fast-forwarded to HEAD of `master`so that recent changes are contained.
I will prepare a new beta version replacing 2022.01.02 tonight.

@sk-juroot and others:
Please from now on target all PRs with fixes which **shall be** included before beta goes into production to branch `release`.
Bigger changes or changes needing more test and **shall not be** contained in the next release shall be targeted to `master`.

See https://github.com/cgeo/WhereYouGo#branches for branch use explanation.
The principle is the same as described for c:geo here: https://github.com/cgeo/cgeo/wiki/How-to-get-a-bug-fix-into-the-release

**Comment by Lineflyer** [comment]
Beta 2022.01.10 has been pushed to beta testers on Google Play and is currently digested/checked by Google Play.

**Comment by Lineflyer** [comment]
I have issued the following posting on Facebook and also sent it to allow beta users previously contacting us on support mail. IMHO this is the only way to get on the "preserve train" again:

Thanks to your feedback we were able to identify a major problem with file access in the current beta version.
On newer Android versions the first WhereYouGo beta version (2021.01.02) is no longer able to access files from random directories (like custom cartridge directory and offline maps).

Unfortunately this problem can`t be undone by another update due to Google file access restrictions. We are really sorry for that trouble, but there is a lot still to be done to modernize WhereYouGo.

Beta users, which have the above described problems should do the following steps:

- Leave the beta program by opting out on https://play.google.com/apps/testing/menion.android.whereyougo
- Uninstall the beta version and wait some time to allow Google to remove you from beta
- Install the normal version (needs to be version 2020.08.21) from Google Play once available and set up as needed
- Join the beta program again on https://play.google.com/apps/testing/menion.android.whereyougo
- Once a new beta version is announced, please double check its version 2022.01.10 and update to it

This new beta version should keep file access available but only if you are upgrading from the previous version.
Sorry for the inconvenience and please share to all WhereYouGo beta users!

**Comment by Lineflyer** [comment]
The newer beta version 2022.01.10 seems to perform good. We even got some confirmations, that the above described scenario reenables file access to upgrading users.

So the best we can do for now seems to be to publish the current beta.

This will fix some bugs and enable Android 12 installations to use GPS again. 
The downside however is, that new installations will be affected by following limitations/issues:
- No offline mspa available
- Cartridges only storable in default directory and only via in-app download

Given the fact, that there are almost no new-installations at the moment but a larger audience with Android 12 problem, this is probably the way to go for now, hoping that we get scoped storage into WhereYouGo next months.

Will publish on the weekend if no one objects.

Side info:
The publish will be combined with a legacy release for API16-20, see https://github.com/cgeo/WhereYouGo/issues/333#issuecomment-1012546530

**Comment by ztNFny** [comment]
> Cartridges only storable in default directory and only via in-app download

Maybe an "import from sdcard" can be added as  a "cheap"workaround for this?

**Comment by Lineflyer** [comment]
I think not, because we cannot access any files anymore.

**Comment by Lineflyer** [comment]
Release ongoing for 2022.01.14 (Android 5+) and 2022.01.14-legacy (Android 4.1 to 4.4):

- [x] Check/Insert correct version name as headline in `/src/main/assets/news.xml`.
- [x] Check/Complete the versions changelog in `src/main/assets/news.xml` (will be shown in the app) 
- [x] Build release on CI server
- [ ] Perform smoke test with resulting ```whereyougo-release.apk``` before continuing
- [ ] Upload and publish ```whereyougo-release.apk``` to Google Play
- [ ] Tag the release on Github  (this will also trigger a new release on F-Droid)
- [ ] Publish the release on Github including upload of ```whereyougo-release.apk``` to Github
- [ ] Post release info to Facebook as soon as update is available on Google Play

**Comment by Lineflyer** [comment]
Google Play is now digesting:
- 2022.01.14 (20220114) for Android 5.0 and above
- 2022.01.14-legacy (20220105) for Android 4.1 to 4.4

Will close once available.

Once again, the limitations known are:
If you do not come from the upgrade path 2020.08.21 (or earlier) to this version, there will be no file access. This means, no offline maps and cartridges only via in-app download to default directory.
Needs to be solved by fixing #309

**Comment by Lineflyer** [comment]
Release is done. Now proper file access should be highest priority.
